Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,514 @@
+        if exponentPartRange.location != NSNotFound, let range = Range(exponentPartRange, in: data) {
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if let pointIndex = decimalPart.firstIndex(of: ".") {
+            // move the exponent by the number of digits after the decimal point
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.maxSignificandDigits {
+            // Clamping upper bound: Exponent is too large, try shifting zeros into the coefficient
+
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Clamping lower bound: Exponent is too small, try taking zeros off the coefficient
+
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error")
+            }
+        }
+
+        guard (Self.exponentMin...Self.exponentMax).contains(exponent) else {
+            throw BSONError.InvalidArgumentError(
+                message: "Exponent \(exponent) is out of encodable range \(Self.exponentMin...Self.exponentMax)"
+            )
+        }
+
+        guard digits.count <= Self.maxSignificandDigits else {
+            throw BSONError.InvalidArgumentError(message: "Overflow Error")
+        }
+
+        let significandLoDigits = [UInt8](digits.suffix(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+        let significandHiDigits = [UInt8](digits.dropLast(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+
+        // Multiply by one hundred quadrillion (note the seventeen zeroes)
+        // the product is the significandHiDigits "shifted" up by 17 decimal places
+        // we can then add the significandLoDigits to the product to ensure that we have a correctly formed significand
+        var significand = UInt128.multiply(significandHiDigits, by: Self.decimalShift17Zeroes)
+
+        let (result, didOverflow) = significand.lo.addingReportingOverflow(significandLoDigits)
+        significand.lo = result
+
+        if didOverflow {
+            significand.hi += 1
+        }
+
+        let biasedExponent = UInt64(exponent + Self.exponentBias).getLeastSignificantBits(Self.exponentLength)
+
+        var value = UInt128()
+
+        value.hi |= biasedExponent << (63 - Self.exponentLength)
+        value.hi |= significand.hi.getLeastSignificantBits(60)
```

Comment:
I think this should be 48 (63 - 1 - 14) since the significand is only 113 bits (and also to make room for the exponent). I don't think it should actually matter since the higher order bits will all be 0 though, so maybe it just makes sense to do `value.hi |= significand.hi`.